File: packages/javascript/tests/rendering/asset-binding-contract.test.ts
import { describe, expect, it } from 'vitest';
import {
  PatchMapAssetRuntime,
  PATCH_MAP_DEFAULT_ASSET_POLICY,
  PATCH_MAP_BUILTIN_ASSETS,
  type PatchMapAssetBackend,
  type PatchMapAssetBackendRequest,
} from '../../src';
import { PatchMapAssetError, evaluatePatchMapAssetResponsePolicy } from '../../src/assets';
import { createPatchMapApi } from '../../src/public';
import { createHost } from '../integration/developer-api-host';
import { stableHash64Hex } from '../../src/shared/stable-hash';

function deferred<T>() {
  let resolve!: (value: T | PromiseLike<T>) => void;
  const promise = new Promise<T>((done) => { resolve = done; });
  return { promise, resolve };
}

class ContractBackend implements PatchMapAssetBackend {
  readonly keyNamespace = ' contract ';
  readonly requests: PatchMapAssetBackendRequest[] = [];
  readonly completion = deferred<unknown>();
  unloading: ReturnType<typeof deferred<void>> | undefined;
  failUnload = false;
  get(): unknown { return undefined; }
  load(request: PatchMapAssetBackendRequest): Promise<unknown> {
    this.requests.push(request);
    return this.completion.promise;
  }
  async unload(): Promise<void> {
    await this.unloading?.promise;
    if (this.failUnload) throw new Error('cleanup');
  }
}
const error = (code: string, category: string, retryable: boolean) =>
  ({ code, category, retryable });
const settle = async () => { for (let i = 0; i < 12; i += 1) await Promise.resolve(); };

describe('asset public binding witnesses', () => {
  it('asset descriptor registration and policy fields preserve exact public values', async () => {
    expect(PATCH_MAP_DEFAULT_ASSET_POLICY).toEqual({ maxEncodedBytes: 20971520, maxDecodedWidth: 8192, maxDecodedHeight: 8192 });
    expect(PATCH_MAP_BUILTIN_ASSETS.map((value) => value.alias)).toEqual([
      'object', 'inverter', 'combiner', 'device', 'edge', 'loading', 'warning', 'wifi',
      'FiraCode-300', 'FiraCode-400', 'FiraCode-500', 'FiraCode-600', 'FiraCode-700',
    ]);
    const backend = new ContractBackend();
    const runtime = new PatchMapAssetRuntime(backend);
    const input = { src: ' source ', format: ' svg ', parser: ' svg ', loadParser: ' svg ', data: { resolution: 2 } };
    const descriptor = runtime.sourceEntry(input).descriptor;
    expect(descriptor).toEqual({ src: 'source', format: 'svg', parser: 'svg', loadParser: 'svg', data: { resolution: 2 } });
    expect(runtime.sourceEntry(' source ').descriptor).toEqual({ src: 'source' });
    for (const field of ['src', 'format', 'parser', 'loadParser']) {
      for (const invalid of ['', ' ', 4, null]) {
        expect(() => runtime.sourceEntry({ ...descriptor, [field]: invalid } as never))
          .toThrow(expect.objectContaining(error('INVALID_VALUE', 'INVALID_INPUT', false)));
      }
    }
    expect(runtime.registerAssets([
      { alias: ' font ', descriptor, kind: 'font', fontWeight: 500 },
      { alias: 'font', descriptor, kind: 'font', fontWeight: 500 },
      { alias: 'image', descriptor: 'image' },
    ])).toEqual({ registeredAliases: ['font', 'image'], duplicateAliases: ['font'] });
    const font = runtime.resolve('font');
    expect([font.alias, font.descriptor, font.kind, font.fontWeight, font.packageOwned])
      .toEqual(['font', descriptor, 'font', 500, false]);
    expect(runtime.resolve('image').kind).toBe('image');
    for (const invalid of [{ kind: 'other' }, { fontWeight: 0 }, { fontWeight: 2.5 }, { fontWeight: '500' }]) {
      expect(() => runtime.registerAlias({ alias: 'bad', descriptor: 'bad', ...invalid } as never))
        .toThrow(expect.objectContaining(error('INVALID_VALUE', 'INVALID_INPUT', false)));
    }
    expect(() => runtime.registerAlias({ alias: 'font', descriptor: 'conflict' }))
      .toThrow(expect.objectContaining(error('CONFLICT', 'CONFLICT', false)));
    const policy = { maxEncodedBytes: 10, maxDecodedWidth: 20, maxDecodedHeight: 30 };
    for (const field of Object.keys(policy)) {
      for (const invalid of [0, -1, 1.5, Infinity, '1']) {
        expect(() => runtime.createSession({ instanceId: field, policy: { [field]: invalid } as never }))
          .toThrow(expect.objectContaining(error('INVALID_VALUE', 'INVALID_INPUT', false)));
      }
    }
    for (const [field, limit] of [['encodedBytes', 10], ['decodedWidth', 20], ['decodedHeight', 30]] as const) {
      const metadata = { mediaType: 'image/png', encodedBytes: 10, decodedWidth: 20, decodedHeight: 30 };
      expect(evaluatePatchMapAssetResponsePolicy(policy, metadata).accepted).toBe(true);
      expect(evaluatePatchMapAssetResponsePolicy(policy, { ...metadata, [field]: limit + 1 })).toEqual({
        accepted: false, code: 'ASSET_POLICY_REJECTED', stage: field === 'encodedBytes' ? 'encoded-bytes' : 'decoded-size',
      });
    }
    for (const code of ['ASSET_LOAD_FAILED', 'ASSET_DECODE_FAILED', 'ASSET_UPLOAD_FAILED'] as const) {
      const value = new PatchMapAssetError(code, 'ASSET_FAILURE', true);
      expect([value.code, value.category, value.retryable]).toEqual([code, 'ASSET_FAILURE', true]);
    }
    const session = runtime.createSession({ instanceId: 'policy', policy });
    const acquiring = session.acquireSource('source');
    backend.completion.resolve({});
    await acquiring;
    expect(backend.requests[0]?.policy).toEqual(policy);
    await session.destroy();
  });

  it('asset backend requests and runtime session status enumerate lifecycle fields', async () => {
    const backend = new ContractBackend(), runtime = new PatchMapAssetRuntime(backend);
    runtime.registerAlias({ alias: 'image', descriptor: 'source' });
    const identity = runtime.resolve('image').cacheIdentity;
    const policy = { maxEncodedBytes: 10, maxDecodedWidth: 20, maxDecodedHeight: 30 };
    const session = runtime.createSession({ instanceId: ' instance ', policy });
    const resource = (state: string, resources = 1, pending = 0, leases = 0, ownership: string | null = 'patch-map', cleanup = false) => ({
      cacheIdentity: identity, resourceCount: resources, pendingCount: pending, leaseCount: leases,
      ownership, state, cleanupPending: cleanup, cleanupRetryOwner: cleanup ? 'runtime' : null,
    });
    expect(runtime.probe('image').resource).toEqual(resource('absent', 0, 0, 0, null));
    expect(runtime.probe().resource).toBeNull();
    const pending = session.acquire('image');
    await settle();
    expect(runtime.probe('image')).toEqual({
      builtins: { aliases: ['object', 'inverter', 'combiner', 'device', 'edge', 'loading', 'warning', 'wifi'] },
      fonts: { weights: [300, 400, 500, 600, 700] }, aliasCount: 1, resourceCount: 1,
      pendingCount: 1, leaseCount: 0, cleanupPendingCount: 0, resource: resource('pending', 1, 1),
    });
    expect(session.probe()).toEqual({ instanceId: 'instance', destroyed: false, pendingCount: 1,
      leaseCount: 0, acquisitionCount: 1, cleanupPendingCount: 0 });
    const request = backend.requests[0];
    expect(request?.key).toMatch(/^patch-map-asset:/u);
    expect(request?.key.startsWith(`patch-map-asset:${stableHash64Hex('contract')}:`)).toBe(true);
    expect(request?.cacheIdentity).toBe(identity);
    expect(request?.descriptor).toEqual({ src: 'source' });
    expect(request?.packageOwned).toBe(false);
    expect(request?.policy).toEqual(policy);
    backend.completion.resolve({});
    const handle = await pending;
    expect(runtime.probe('image').resource).toEqual(resource('resolved', 1, 0, 1));
    backend.unloading = deferred<void>();
    backend.failUnload = true;
    const release = handle.release();
    const failed = expect(release).rejects.toThrow(expect.objectContaining(error('INTERNAL_FAILURE', 'INTERNAL_FAILURE', false)));
    await settle();
    expect(runtime.probe('image').resource).toEqual(resource('releasing'));
    backend.unloading.resolve();
    await failed;
    expect(runtime.probe('image').resource).toEqual(resource('cleanup-failed', 1, 0, 0, 'patch-map', true));
    expect(session.probe().cleanupPendingCount).toBe(1);
    backend.failUnload = false;
    await runtime.retryCleanup();
    expect(runtime.probe('image').resource).toEqual(resource('absent', 0, 0, 0, null));
    await session.destroy();
    expect(session.probe()).toEqual({ instanceId: 'instance', destroyed: true, pendingCount: 0,
      leaseCount: 0, acquisitionCount: 0, cleanupPendingCount: 0 });
    expect(() => session.acquire('image')).toThrow(expect.objectContaining(error('CANCELLED', 'CANCELLED', false)));
    const facadeSession = runtime.createSession({ instanceId: 'public' });
    let present = true;
    const api = createPatchMapApi({
      ...createHost().host,
      registerAssets: (_instanceId, registrations) => runtime.registerAssets(registrations),
      assetProbe: (alias) => ({ runtime: runtime.probe(alias), session: present ? facadeSession.probe() : null }),
    });
    expect(api.assets.register({ alias: 'public', descriptor: 'public' }))
      .toEqual({ registeredAliases: ['public'], duplicateAliases: [] });
    expect(api.assets.register([{ alias: 'public', descriptor: 'public' }]))
      .toEqual({ registeredAliases: [], duplicateAliases: ['public'] });
    expect(api.assets.status('public')).toEqual({ runtime: runtime.probe('public'), session: facadeSession.probe() });
    present = false;
    expect(api.assets.status().session).toBeNull();
    await facadeSession.destroy();
  });
});

File: packages/flutter/test/host/asset_binding_contract_test.dart
import 'dart:async';
import 'package:flutter_test/flutter_test.dart';
import 'package:conalog_patch_map/conalog_patch_map.dart';
import 'package:conalog_patch_map/src/host/native_assets.dart' show NativeAssetSession;

class ContractBackend extends PatchMapAssetBackend {
  final completion = Completer<Object?>();
  final requests = <PatchMapAssetBackendRequest>[];
  Completer<void>? unloading;
  bool failUnload = false;
  @override
  String get keyNamespace => ' contract ';
  @override
  Object? get(PatchMapAssetBackendRequest request) => null;
  @override
  Future<Object?> load(PatchMapAssetBackendRequest request) {
    requests.add(request);
    return completion.future;
  }

  @override
  Future<void> unload(String key) async {
    await unloading?.future;
    if (failUnload) throw StateError('cleanup');
  }
}

Matcher assetError(String code, String category, bool retryable) =>
    isA<PatchMapAssetError>()
        .having((e) => e.code, 'code', code)
        .having((e) => e.category, 'category', category)
        .having((e) => e.retryable, 'retryable', retryable);

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();
  test(
    'public native backend factory decodes and releases an actual SVG source',
    () async {
      final backend = createPatchMapNativeAssetBackend();
      final request = PatchMapAssetBackendRequest(
        key: 'factory-svg',
        cacheIdentity: 'fixture-svg',
        packageOwned: false,
        policy: normalizePatchMapAssetPolicy(),
        descriptor: {
          'src': Uri.dataFromString(
            '<svg xmlns="http://www.w3.org/2000/svg" width="2" height="3"><rect width="2" height="3" fill="red"/></svg>',
            mimeType: 'image/svg+xml',
          ).toString(),
        },
      );
      expect(backend.get(request), null);
      final asset = await backend.load(request) as NativeAsset;
      expect([asset.width, asset.height, asset.isFont], [2, 3, false]);
      expect(asset.picture, isNotNull);
      await backend.unload(request.key);
      await backend.unload(request.key);
      asset.dispose();
    },
  );
  test(
    'asset descriptor registration and policy fields preserve exact public values',
    () {
      expect(patchMapDefaultAssetPolicy, {
        'maxEncodedBytes': 20971520,
        'maxDecodedWidth': 8192,
        'maxDecodedHeight': 8192,
      });
      expect(patchMapBuiltinAssets.map((value) => value['alias']).toList(), [
        'object',
        'inverter',
        'combiner',
        'device',
        'edge',
        'loading',
        'warning',
        'wifi',
        'FiraCode-300',
        'FiraCode-400',
        'FiraCode-500',
        'FiraCode-600',
        'FiraCode-700',
      ]);
      final descriptor = normalizePatchMapAssetDescriptor({
        'src': ' source ',
        'format': ' svg ',
        'parser': ' svg ',
        'loadParser': ' svg ',
        'data': {'resolution': 2},
      });
      expect(descriptor, {
        'src': 'source',
        'format': 'svg',
        'parser': 'svg',
        'loadParser': 'svg',
        'data': {'resolution': 2},
      });
      expect(normalizePatchMapAssetDescriptor(' source '), {'src': 'source'});
      for (final field in ['src', 'format', 'parser', 'loadParser']) {
        for (final invalid in ['', ' ', 4, null]) {
          expect(
            () => normalizePatchMapAssetDescriptor({
              ...descriptor,
              field: invalid,
            }),
            throwsA(assetError('INVALID_VALUE', 'INVALID_INPUT', false)),
          );
        }
      }
      final runtime = PatchMapAssetRuntime(ContractBackend());
      expect(
        runtime.registerAssets([
          {
            'alias': ' font ',
            'descriptor': descriptor,
            'kind': 'font',
            'fontWeight': 500,
          },
          {
            'alias': 'font',
            'descriptor': descriptor,
            'kind': 'font',
            'fontWeight': 500,
          },
          {'alias': 'image', 'descriptor': 'image'},
        ]).toJson(),
        {
          'registeredAliases': ['font', 'image'],
          'duplicateAliases': ['font'],
        },
      );
      final font = runtime.resolve('font');
      expect(
        [
          font.alias,
          font.descriptor,
          font.kind,
          font.fontWeight,
          font.packageOwned,
        ],
        ['font', descriptor, 'font', 500, false],
      );
      expect(runtime.resolve('image').kind, 'image');
      for (final code in [
        'ASSET_LOAD_FAILED',
        'ASSET_DECODE_FAILED',
        'ASSET_UPLOAD_FAILED',
      ]) {
        final value = PatchMapAssetError(code, 'ASSET_FAILURE', true);
        expect(
          [value.code, value.category, value.retryable],
          [code, 'ASSET_FAILURE', true],
        );
      }
      for (final invalid in [
        {'kind': 'other'},
        {'fontWeight': 0},
        {'fontWeight': 2.5},
        {'fontWeight': '500'},
      ]) {
        expect(
          () => runtime.registerAlias({
            'alias': 'bad',
            'descriptor': 'bad',
            ...invalid,
          }),
          throwsA(assetError('INVALID_VALUE', 'INVALID_INPUT', false)),
        );
      }
      expect(
        () =>
            runtime.registerAlias({'alias': 'font', 'descriptor': 'conflict'}),
        throwsA(assetError('CONFLICT', 'CONFLICT', false)),
      );
      const policy = {
        'maxEncodedBytes': 10,
        'maxDecodedWidth': 20,
        'maxDecodedHeight': 30,
      };
      expect(normalizePatchMapAssetPolicy(policy), policy);
      expect(normalizePatchMapAssetPolicy(), patchMapDefaultAssetPolicy);
      for (final field in policy.keys) {
        for (final invalid in [0, -1, 1.5, double.infinity, '1']) {
          expect(
            () => normalizePatchMapAssetPolicy({field: invalid}),
            throwsA(assetError('INVALID_VALUE', 'INVALID_INPUT', false)),
          );
        }
        final metadata = <String, dynamic>{
          'mediaType': 'image/png',
          'encodedBytes': 10,
          'decodedWidth': 20,
          'decodedHeight': 30,
        };
        expect(
          evaluatePatchMapAssetResponsePolicy(policy, metadata)['accepted'],
          true,
        );
        final key = {
          'maxEncodedBytes': 'encodedBytes',
          'maxDecodedWidth': 'decodedWidth',
          'maxDecodedHeight': 'decodedHeight',
        }[field]!;
        metadata[key] = policy[field]! + 1;
        expect(evaluatePatchMapAssetResponsePolicy(policy, metadata), {
          'accepted': false,
          'code': 'ASSET_POLICY_REJECTED',
          'stage': field == 'maxEncodedBytes'
              ? 'encoded-bytes'
              : 'decoded-size',
        });
        expect(
          () => assertPatchMapAssetResponseAllowed(policy, metadata),
          throwsA(assetError('ASSET_POLICY_REJECTED', 'ASSET_FAILURE', false)),
        );
      }
    },
  );

  test(
    'asset backend requests and runtime session status enumerate lifecycle fields',
    () async {
      final backend = ContractBackend(),
          runtime = PatchMapAssetRuntime(ContractBackend());
      final shared = PatchMapAssetRuntime(backend);
      shared.registerAlias({'alias': 'image', 'descriptor': 'source'});
      final identity = shared.resolve('image').cacheIdentity;
      expect(backend.keyNamespace, ' contract ');
      const policy = {
        'maxEncodedBytes': 10,
        'maxDecodedWidth': 20,
        'maxDecodedHeight': 30,
      };
      final session = shared.createSession(
        instanceId: ' instance ',
        policy: policy,
      );
      Map<String, dynamic> resource(
        String state, {
        int resources = 1,
        int pending = 0,
        int leases = 0,
        String? ownership = 'patch-map',
        bool cleanup = false,
      }) => {
        'cacheIdentity': identity,
        'resourceCount': resources,
        'pendingCount': pending,
        'leaseCount': leases,
        'ownership': ownership,
        'state': state,
        'cleanupPending': cleanup,
        'cleanupRetryOwner': cleanup ? 'runtime' : null,
      };
      expect(
        shared.probe('image')['resource'],
        resource('absent', resources: 0, ownership: null),
      );
      expect(runtime.probe()['resource'], null);
      final pending = session.acquire('image');
      await Future<void>.delayed(Duration.zero);
      expect(shared.probe('image'), {
        'builtins': {
          'aliases': [
            'object',
            'inverter',
            'combiner',
            'device',
            'edge',
            'loading',
            'warning',
            'wifi',
          ],
        },
        'fonts': {
          'weights': [300, 400, 500, 600, 700],
        },
        'aliasCount': 1,
        'resourceCount': 1,
        'pendingCount': 1,
        'leaseCount': 0,
        'cleanupPendingCount': 0,
        'resource': resource('pending', pending: 1),
      });
      expect(session.probe(), {
        'instanceId': 'instance',
        'destroyed': false,
        'pendingCount': 1,
        'leaseCount': 0,
        'acquisitionCount': 1,
        'cleanupPendingCount': 0,
      });
      final request = backend.requests.single;
      expect(request.key, startsWith('contract:'));
      expect(request.cacheIdentity, identity);
      expect(request.descriptor, {'src': 'source'});
      expect(request.packageOwned, false);
      expect(request.policy, policy);
      backend.completion.complete(Object());
      final handle = await pending;
      expect(
        shared.probe('image')['resource'],
        resource('resolved', leases: 1),
      );
      backend.unloading = Completer<void>();
      backend.failUnload = true;
      final release = handle.release();
      final failed = expectLater(
        release,
        throwsA(assetError('INTERNAL_FAILURE', 'INTERNAL_FAILURE', false)),
      );
      await Future<void>.delayed(Duration.zero);
      expect(shared.probe('image')['resource'], resource('releasing'));
      backend.unloading!.complete();
      await failed;
      expect(
        shared.probe('image')['resource'],
        resource('cleanup-failed', cleanup: true),
      );
      expect(session.probe()['cleanupPendingCount'], 1);
      backend.failUnload = false;
      await shared.retryCleanup();
      expect(
        shared.probe('image')['resource'],
        resource('absent', resources: 0, ownership: null),
      );
      await session.destroy();
      expect(session.probe(), {
        'instanceId': 'instance',
        'destroyed': true,
        'pendingCount': 0,
        'leaseCount': 0,
        'acquisitionCount': 0,
        'cleanupPendingCount': 0,
      });
      expect(
        () => session.acquire('image'),
        throwsA(assetError('CANCELLED', 'CANCELLED', false)),
      );
      final host = NativeAssetSession(
        runtime: runtime,
        instanceId: 'host',
        fontInitializer: () async {},
      );
      final controller = await PatchMapController.create(assetPort: host);
      expect(
        controller.assets.register({
          'alias': 'public',
          'descriptor': 'public',
        }).toJson(),
        {
          'registeredAliases': ['public'],
          'duplicateAliases': [],
        },
      );
      expect(
        controller.assets.register([
          {'alias': 'public', 'descriptor': 'public'},
        ]).toJson(),
        {
          'registeredAliases': [],
          'duplicateAliases': ['public'],
        },
      );
      final status = controller.assets.status('public');
      expect(status.keys, ['session', 'runtime']);
      expect(status['runtime'], runtime.probe('public'));
      expect(status['session'], {
        'instanceId': 'host',
        'destroyed': false,
        'pendingCount': 0,
        'leaseCount': 0,
        'acquisitionCount': 0,
        'cleanupPendingCount': 0,
        'resolved': 0,
        'failed': 0,
        'ready': false,
      });
      await controller.destroy();
      expect(controller.assets.status(), {
        'session': null,
        'runtime': runtime.probe(),
      });
    },
  );
}
